proof_of_leadership/leadership_poseidon.circom

```diff
@@ -0,0 +1,301 @@
+include "poseidon/poseidon_4_to_1_Jubjub.circom";
+include "poseidon/poseidon_16_to_1_Jubjub.circom";
+include "../../circomlib-master/circuits/bitify.circom";
+
+template BLSLessThan(n) {
+    assert(n <= 253);
+    signal input in[2];
+    signal output out;
+
+    component n2b = Num2Bits(n+1);
+
+    n2b.in <== in[0]+ (1<<n) - in[1];
+
+    out <== 1-n2b.out[n];
+}
+
+template BLSNum2Bits_strict() {
+    signal input in;
+    signal output out[255];
+
+        // Ensure that out is lower than p
+    component check_range = CompConstant(23487852865797141623554994256013988874373056334117496812739262697960298774528); // -1 - 2**254 (p-1 without its first bit)
+    component n2b = Num2Bits(255);
+    in ==> n2b.in;
+
+    for (var i=0; i<255; i++) {
+        n2b.out[i] ==> out[i];
+        if(i != 0){
+            n2b.out[i] ==> check_range.in[i-1];
+        }
+    }
+
+    check_range.out * (n2b.out[0]) === 0; //must be zero exept if the first bit is 0 => then in is on 254 bits and p-1 on 255
+}
+
+template BLSBits2Num_strict() {
+    signal input in[255];
+    signal output out;
+            //ensure that in is not greater than p
+    component check_range = CompConstant(23487852865797141623554994256013988874373056334117496812739262697960298774528);
+    component b2n = Bits2Num(255);
+
+    for (var i=0; i<255; i++) {
+        in[i] ==> b2n.in[i];
+        if(i != 0){
+            in[i] ==> check_range.in[i-1];
+        }
+    }
+
+    check_range.out * in[0] === 0;
+
+    b2n.out ==> out;
+}
+
+template check_bits(n){
+    signal input bits[n];
+    for(var i=0; i<n; i++){
+        bits[i] * (1-bits[i]) === 0;
+    }
+}
+
+template check_lottery(){
+    signal input epoch_nonce;
+    signal input slot_number;
+    signal input t0;
+    signal input t1;        // The precomputed threshold values
+
+    signal input constraints;
+    signal input value;
+    signal input unit;
+    signal input state;
+    signal input note_nonce;
+    signal input nullifier_secret_key;
+    signal input randomness;
+
+    component hash = hash_16_to_1();
+
+            //The b"lead" Tag converted in F_p element (from bits with big endian order)
+    hash.in[0] <== 1818583396;
+
+    hash.in[1] <== epoch_nonce;
+    hash.in[2] <== slot_number;
+    hash.in[3] <== constraints;
+    hash.in[4] <== value;
+    hash.in[5] <== unit;
+    hash.in[6] <== state;
+    hash.in[7] <== note_nonce;
+    hash.in[8] <== nullifier_secret_key;
+    hash.in[9] <== randomness;
+    for(var i=10; i<16; i++){
+        hash.in[i] <== 0;
+    }
+
+            // As p-1 is divisible by 4, if X follows a uniform distribution between 0 and p-1
+            // Y = X % (p-1)/4 nearly follows a uniform distribution between 0 and ((p-1)/4)-1 (exept that 0 has 20% more chance to appear than another element which is negligible)
+            // So we transform the hash of a maximum of 255 to a hash of a maximum of 253 bits by taking its modulo
```

Comment:
Why do we need to reduce to 253 bits here? I recall you mentioning that we needed this for poseidon but I forget why

Reply:
First p is 255 bit.
Since p isn't a power of 2 and to avoid overflow, we drop to 254 bits.
We need 1 bit for the sign for comparison, =>253 bits.

Reply:
If we switch to a different curve, all the parameters must change including this.